service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/ResourceExtractor.java

```diff
@@ -2,19 +2,63 @@
 
 import java.util.Optional;
 
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
 import org.hl7.fhir.dstu3.model.Bundle;
+import org.hl7.fhir.dstu3.model.ListResource;
 import org.hl7.fhir.dstu3.model.Resource;
+import org.hl7.fhir.dstu3.model.ResourceType;
 import org.hl7.fhir.instance.model.api.IIdType;
 
 public class ResourceExtractor {
 
     public static Optional<Resource> extractResourceByReference(Bundle allBundle, IIdType reference) {
-        return allBundle.getEntry()
-            .stream()
-            .filter(Bundle.BundleEntryComponent::hasResource)
-            .map(Bundle.BundleEntryComponent::getResource)
-            .filter(resource -> resource.getResourceType().name().equals(reference.getResourceType()))
-            .filter(resource -> resource.getIdElement().getIdPart().equals(reference.getIdPart()))
-            .findFirst();
+        if (ObjectUtils.allNotNull(allBundle, reference)) {
+            return allBundle.getEntry()
+                .stream()
+                .filter(Bundle.BundleEntryComponent::hasResource)
+                .map(Bundle.BundleEntryComponent::getResource)
+                .filter(resource -> isReferencedResource(reference, resource))
+                .findFirst();
+        }
+
+        return Optional.empty();
+    }
+
+    private static boolean isReferencedResource(IIdType reference, Resource resource) {
+        return hasResourceType(resource, reference.getResourceType()) && hasId(reference, resource);
+    }
+
+    public static Optional<ListResource> extractListByEncounterReference(Bundle allBundle, IIdType reference, String code) {
+        if (ObjectUtils.allNotNull(allBundle, reference, code)) {
+            return allBundle.getEntry()
+                .stream()
+                .filter(Bundle.BundleEntryComponent::hasResource)
+                .map(Bundle.BundleEntryComponent::getResource)
+                .filter(resource -> hasResourceType(resource, ResourceType.List.name()))
+                .map(resource -> (ListResource) resource)
+                .filter(resource -> hasEncounterReference(reference, resource) && hasCode(code, resource))
+                .findFirst();
+        }
+
+        return Optional.empty();
+    }
+
+    private static boolean hasResourceType(Resource resource, String resourceType) {
+        return StringUtils.equals(resource.getResourceType().name(), resourceType);
+    }
+
+    private static boolean hasId(IIdType reference, Resource resource) {
+        return resource.hasIdElement() && StringUtils.equals(resource.getIdElement().getIdPart(), reference.getIdPart());
+    }
+
+    private static boolean hasEncounterReference(IIdType reference, ListResource resource) {
+        return resource.hasEncounter()
+            && resource.getEncounter().hasReferenceElement()
+            && StringUtils.equals(resource.getEncounter().getReferenceElement().getIdPart(), reference.getIdPart());
```

Comment:
Could we just use "equals" String object method?

Reply:
This utils also compares nulls, and due to that there is no need to check if in this case idPart is not null before calling equal on it, so preventing NullPointerException. I think it's better to use those string utils to compare strings.